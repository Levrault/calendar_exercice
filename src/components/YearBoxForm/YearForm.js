import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { reduxForm } from 'redux-form';
import styles from './YearForm.css';
import CountDownField from '../Field/CountDownField';

/**
 * Create a profile
 */
class YearForm extends Component {
  /**
   * onSubmit
   */
  onSubmit = (values) => {
    this.props.handleSubmit(values);
  }

  /**
  * Render
  */
  render () {
    return (
      <form className={styles.profileForm} onSubmit={this.onSubmit}>
        <CountDownField name="year" />
      </form>
    );
  }
}

YearForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired
};

export default reduxForm({
  form: 'year'
})(YearForm);
