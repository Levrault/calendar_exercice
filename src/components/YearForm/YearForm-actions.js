import axios from 'axios';
import minBy from 'lodash/minBy';
import maxBy from 'lodash/maxBy';
import { eventsInit } from '../Event/Events-actions';

export const YEAR_FETCH_BEGIN = 'CALENDAR_FETCH_BEGIN';
export const YEAR_FETCH_SUCCESS = 'CALENDAR_FETCH_SUCCESS';
export const YEAR_FETCH_FAILURE = 'CALENDAR_FETCH_FAILURE';
export const YEAR_FETCH_ALL_BEGIN = 'CALENDAR_FETCH_ALL_BEGIN';
export const YEAR_FETCH_ALL_SUCCESS = 'CALENDAR_FETCH_ALL_SUCCESS';
export const YEAR_FETCH_ALL_FAILURE = 'CALENDAR_FETCH_ALL_FAILURE';

/**
 * Begin to fetch data
 * @returns {object}
 */
export const yearFetchAllBegin = () => ({
  type: YEAR_FETCH_ALL_BEGIN
});

/**
 * Calendar data has been returned
 * @param {array} months
 * @returns {object}
 */
export const yearFetchAllSuccess = data => ({
  type: YEAR_FETCH_ALL_SUCCESS,
  payload: {
    data,
    min: minBy(data, 'year').year,
    max: maxBy(data, 'year').year
  }
});

/**
 * Receive an error
 * @param {object} error
 * @returns {object}
 */
export const yearFetchAllError = error => ({
  type: YEAR_FETCH_ALL_FAILURE,
  payload: { error }
});

/**
 * Begin to fetch data
 * @returns {object}
 */
export const yearFetchBegin = () => ({
  type: YEAR_FETCH_BEGIN
});

/**
 * Calendar data has been returned
 * @param {array} current
 * @returns {object}
 */
export const yearFetchSuccess = current => ({
  type: YEAR_FETCH_SUCCESS,
  payload: {
    current
  }
});

/**
 * Receive an error
 * @param {object} error
 * @returns {object}
 */
export const yearFetchError = error => ({
  type: YEAR_FETCH_FAILURE,
  payload: { error }
});

/**
 *  Get a specefic year with all his data
 * @param {string} id
 */
export const fetch = (id) => async (dispatch) => {
  dispatch(yearFetchBegin());
  try {
    const { data } = await axios.get(`${BASE_URL}calendar/${id}`, {
      params: {
        full: true
      }
    });
    dispatch(yearFetchSuccess(data));
    dispatch(eventsInit(data));
  } catch (error) {
    dispatch(yearFetchError(error));
  }
};

/**
 * Fetch all year ids
 * @param {bool} full
 */
export const fetchAll = (full) => async (dispatch) => {
  dispatch(yearFetchAllBegin());
  try {
    const { data } = await axios.get(`${BASE_URL}calendar`, {
      params: {
        full
      }
    });
    dispatch(yearFetchAllSuccess(data));
  } catch (error) {
    dispatch(yearFetchAllError(error));
  }
};
