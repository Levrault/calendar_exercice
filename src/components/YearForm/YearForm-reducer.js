import {
  YEAR_FETCH_ALL_BEGIN,
  YEAR_FETCH_ALL_SUCCESS,
  YEAR_FETCH_ALL_FAILURE,
  YEAR_FETCH_BEGIN,
  YEAR_FETCH_SUCCESS,
  YEAR_FETCH_FAILURE
} from './YearForm-actions';

const initialState = {
  data: [],
  loading: false,
  error: null,
  current: {},
  min: 2018,
  max: 2025
};

export const yearFormReducer = (state = initialState, action) => {
  switch (action.type) {
  case YEAR_FETCH_ALL_BEGIN:
    return {
      ...state,
      loading: true,
      error: null
    };

  case YEAR_FETCH_ALL_SUCCESS:
    return {
      ...state,
      loading: false,
      ...action.payload
    };

  case YEAR_FETCH_ALL_FAILURE:
    return {
      ...state,
      loading: false,
      error: action.payload.error,
      data: []
    };
  case YEAR_FETCH_BEGIN:
    return {
      ...state,
      loading: true,
      error: null
    };

  case YEAR_FETCH_SUCCESS:
    return {
      ...state,
      loading: false,
      ...action.payload
    };

  case YEAR_FETCH_FAILURE:
    return {
      ...state,
      loading: false,
      error: action.payload.error
    };

  default:
    return state;
  }
};
