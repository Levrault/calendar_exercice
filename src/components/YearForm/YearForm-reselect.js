import { createSelector } from 'reselect';
import get from 'lodash/get';

const yearSelector = state => get(state, 'year');

export const dataSelector = createSelector(
  yearSelector,
  year => get(year, 'data', [])
);

export const monthsSelector = createSelector(
  yearSelector,
  calendar => get(calendar, 'current.months', [])
);

export const currentYearSelector = createSelector(
  yearSelector,
  calendar => get(calendar, 'current.year', 2018)
);

export const currentMonthsSelector = createSelector(
  yearSelector,
  calendar => get(calendar, 'current.year.months', [])
);

export const currentYearIdSelector = createSelector(
  yearSelector,
  calendar => get(calendar, 'current.id', '')
);

export const loadingSelector = createSelector(
  yearSelector,
  calendar => get(calendar, 'loading', false)
);

export const errorSelector = createSelector(
  yearSelector,
  calendar => get(calendar, 'error', null)
);

export const minSelector = createSelector(
  yearSelector,
  calendar => get(calendar, 'min', 2015)
);

export const maxSelector = createSelector(
  yearSelector,
  calendar => get(calendar, 'max', 2025)
);
