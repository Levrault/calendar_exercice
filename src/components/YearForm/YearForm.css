.countdown {
  align-items: center;
  display: flex;
  justify-content: center;
}

.button {
  background: none;
  border: 0;
  color: $palette-secondary-text;
  cursor: pointer;
  margin: 0;
  outline: 0;
  padding: 2px 0 0;

  &.inactive {
    color: $palette-light-primary;
    cursor: inherit;
  }
}

.input {
  border: 0;
  color: $palette-secondary-text;
  font-size: 1.2rem;
  text-align: right;
  width: $space-xl;
}

