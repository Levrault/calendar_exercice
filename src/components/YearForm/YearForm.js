import React, { PureComponent } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import styles from './YearForm.css';

/**
 * Get calendar by year
 * @class YearForm
 * @param {function} fetchAll
 * @param {function} fetch
 * @param {number} min
 * @param {number} max
 * @param {any} ids
 */
class YearForm extends PureComponent {
  /**
   * @constructor
   * @param {object} props
   */
  constructor (props) {
    super(props);
    this.state = {
      value: props.min
    };
  }

  /**
   * fetch data
   */
  componentDidMount () {
    this.props.fetchAll();
  }

  /**
   * Fetch an year
   * @param {object} prevProps
   * @param {object} prevState
   */
  componentDidUpdate (prevProps, prevState) {
    // value is updated
    const { value } = this.state;
    if (value !== prevState.value) {
      this.fetchData(value);
    }

    // first fetching
    if (prevProps.ids !== this.props.ids) {
      this.fetchData(value);
    }
  }

  /**
   * fetch data
   * @param {number} value
   */
  fetchData = (value) => {
    const { ids, fetch } = this.props;
    const year = ids.find(elem => elem.year === value);
    if (year && year.id) {
      fetch(year.id);
    }
  }

  /**
   * @param {number} value
   * @returns {function}
   */
  updateValue = (value) => () => {
    const { min, max } = this.props;

    // if value are in our range
    if (value >= min && value <= max) {
      this.setState({
        value
      });
    }
  }

  /**
   * @returns {node}
   */
  render () {
    const { min, max, ids } = this.props;
    const { value } = this.state;

    // @todo set a real loader
    if (!ids.length) return <span> Loading </span>;

    const leftChevronClassName = classnames(styles.button, {
      [styles.inactive]: value === min
    });
    const rightChevronClassName = classnames(styles.button, {
      [styles.inactive]: value === max
    });
    return (
      <form className={styles.profileForm}>
        <span className={styles.countdown}>

          <button className={leftChevronClassName} type="button" onClick={this.updateValue(value - 1)}>
            <i className="material-icons">chevron_left</i>
          </button>

          <input className={styles.input} type="number" value={value} readOnly />

          <button className={rightChevronClassName} type="button" onClick={this.updateValue(value + 1)}>
            <i className="material-icons">chevron_right</i>
          </button>

        </span>
      </form>
    );
  }
}

YearForm.propTypes = {
  fetchAll: PropTypes.func.isRequired,
  fetch: PropTypes.func.isRequired,
  min: PropTypes.number,
  max: PropTypes.number,
  ids: PropTypes.any
};

YearForm.defaultProps = {
  min: 2018,
  max: 2019
};

export default YearForm;
