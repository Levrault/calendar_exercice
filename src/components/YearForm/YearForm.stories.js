import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, number } from '@storybook/addon-knobs/react';
import { withMarkdownNotes } from '@storybook/addon-notes';
import Wrapper from '../../tools/storybook/Wrapper';
import YearForm from './YearForm';
import getFakeData from '../../tools/test/fakeData';

storiesOf('YearForm', module)
  .addDecorator(withKnobs)
  .addDecorator(Wrapper)
  .add('default', withMarkdownNotes(`
    # YearForm

    Is used to change of calendar. It will get all calendar
    id and get the next calendar available.
    `)(() => (
    <YearForm
      fetchAll={() => {}}
      fetch={() => {}}
      min={number('min', 2018)}
      max={number('max', 2019)}
      ids={getFakeData().year.data}
    />
  )));

