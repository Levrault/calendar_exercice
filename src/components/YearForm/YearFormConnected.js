import { connect } from 'react-redux';
import YearForm from './YearForm';

/**
 * @param {object} state
 * @returns {object}
 */
const mapStateToProps = () => {
  return {};
};

export default connect(mapStateToProps)(YearForm);
