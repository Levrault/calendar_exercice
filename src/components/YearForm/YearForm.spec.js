import React from 'react';
import { shallow, mount } from '../../tools/test/enzyme';
import renderer from 'react-test-renderer';
import YearForm from './YearForm';

const mock = {
  fetchAll: jest.fn(),
  fetch: jest.fn(),
  min: 2018,
  max: 2020,
  ids: [{
    id: '576c020e2bfc5849',
    year: 2018
  },
  {
    id: '576c020e2bfc5810',
    year: 2019
  }]
};

describe('YearForm component', () => {
  beforeEach(() => {
    mock.fetch.mockReset();
    mock.fetchAll.mockReset();
  });

  it('should render without problem', () => {
    const component = renderer.create(
      <YearForm {...mock} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render when loading when getting data', () => {
    const component = renderer.create(
      <YearForm {...mock} ids={[]} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call fetchAll on componentDidMount', () => {
    const spy = jest.spyOn(YearForm.prototype, 'componentDidMount');
    const component = shallow(
      <YearForm {...mock} />
    );
    component.instance().componentDidMount();
    expect(spy).toHaveBeenCalled();
    expect(component.instance().props.fetchAll).toHaveBeenCalled();
    spy.mockRestore();
  });

  it('should call fetch when fetchData is called', () => {
    const component = shallow(
      <YearForm {...mock} />
    );
    component.instance().fetchData(2018);
    expect(component.instance().props.fetch).toHaveBeenCalled();
  });

  it('should not call fetch when value does not exit', () => {
    const component = shallow(
      <YearForm {...mock} />
    );
    component.instance().fetchData(2015);
    expect(component.instance().props.fetch).not.toHaveBeenCalled();
  });

  it('should call fetch when value is changed', () => {
    const component = shallow(
      <YearForm {...mock} />
    );

    component.instance().componentDidUpdate({ ...mock }, { value: 2019 });

    // check new value
    expect(component.instance().props.fetch).toHaveBeenCalled();
  });

  it('should update value when a valid value is passed', () => {
    const component = shallow(
      <YearForm {...mock} />
    );

    // default state
    expect(component.state().value).toBe(mock.min);

    // should update
    const button = component.find('button').last();
    button.simulate('click');

    // check new value
    expect(component.state().value).toBe(2019);
  });

  it('should not update when an invalid value is passed (when lower thant min and highet than max)', () => {
    const component = mount(
      <YearForm {...mock} min={2018} max={2018} />
    );

    // default state
    expect(component.state().value).toBe(mock.min);

    // should update
    const firstButton = component.find('button').first();
    firstButton.simulate('click');

    expect(component.state().value).not.toBe(2017);

    const lastButton = component.find('button').last();
    lastButton.simulate('click');

    expect(component.state().value).not.toBe(2019);
  });

  it('should fetch data only when value change', () => {
    const spy = jest.spyOn(YearForm.prototype, 'componentDidUpdate');
    const component = mount(
      <YearForm {...mock} />
    );

    // default state
    expect(component.state().value).toBe(mock.min);

    // good one
    component.instance().setState({
      value: 2019
    });

    expect(spy).toHaveBeenCalled();
    expect(component.instance().props.fetch).toHaveBeenCalled();
    spy.mockRestore();
  });

  it('should fetch data only when ids have been fetched first', () => {
    const spy = jest.spyOn(YearForm.prototype, 'componentDidUpdate');
    const component = mount(
      <YearForm {...mock} ids={[]} />
    );

    // default state
    expect(component.props().ids.length).toBe(0);

    component.setProps({
      ids: mock.ids
    });

    expect(spy).toHaveBeenCalled();
    expect(component.instance().props.fetch).toHaveBeenCalled();
    spy.mockRestore();
  });
});
