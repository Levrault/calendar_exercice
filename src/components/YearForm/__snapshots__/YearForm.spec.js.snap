// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YearForm component should render when loading when getting data 1`] = `
<span>
   Loading 
</span>
`;

exports[`YearForm component should render without problem 1`] = `
<form>
  <span>
    <button
      className="undefined"
      onClick={[Function]}
      type="button"
    >
      <i
        className="material-icons"
      >
        chevron_left
      </i>
    </button>
    <input
      readOnly={true}
      type="number"
      value={2018}
    />
    <button
      className=""
      onClick={[Function]}
      type="button"
    >
      <i
        className="material-icons"
      >
        chevron_right
      </i>
    </button>
  </span>
</form>
`;
